mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,196 @@
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+from pytools.obj_array import (
+    obj_array_vectorize,
+    obj_array_vectorize_n_args,
+)
+from meshmode.dof_array import thaw
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+from grudge.eager import interior_trace_pair, cross_rank_trace_pairs
+from grudge.symbolic.primitives import TracePair
+from mirgecom.tag_cells import smoothness_indicator
+
+
+def _facial_flux_r(discr, q_tpair):
+
+    actx = q_tpair.int.array_context
+
+    flux_dis = q_tpair.avg
+
+    normal = thaw(actx, discr.normal(q_tpair.dd))
+
+    flux_out = flux_dis * normal
+
+    return discr.project(q_tpair.dd, "all_faces", flux_out)
+
+
+def _facial_flux_q(discr, q_tpair):
+
+    actx = q_tpair[0].int.array_context
+
+    normal = thaw(actx, discr.normal(q_tpair.dd))
+
+    flux_out = np.dot(q_tpair.avg, normal)
+
+    return discr.project(q_tpair.dd, "all_faces", flux_out)
+
+
+def artificial_viscosity(discr, t, eos, boundaries, r, alpha, **kwargs):
+    r"""Compute artifical viscosity for the euler equations.
+
+    Calculates
+    ----------
+    numpy.ndarray
+        The right-hand-side for artificial viscosity for the euler equations.
+
+        .. math::
+
+            \dot{\nabla\cdot{\varepsilon\nabla\mathbf{R}}}
+
+    Parameters
+    ----------
+    r
+        State array which expects the quantity to be limited on to be listed
+        first in the array. For the Euler equations this could be the canonical
+        conserved variables (mass, energy, mometum) for the fluid along with a
+        vector of species masses for multi-component fluids.
+
+    boundaries
+        Dicitionary of boundary functions, one for each valid boundary tag
+
+    t
+        Time
+
+    alpha
+       The maximum artifical viscosity coeffiecent to be applied
+
+    eos: mirgecom.eos.GasEOS
+       Only used as a pass through to the boundary conditions.
+
+    Returns
+    -------
+    numpy.ndarray
+        Agglomerated object array of DOF Arrays representing the RHS associated
+        with the artificial viscosity application.
+    """
+    # Get smoothness indicator
+    indicator = smoothness_indicator(discr, r[0], **kwargs)
+
+    dflux_r = obj_array_vectorize(discr.weak_grad, r)
+
+    def my_facialflux_r_interior(q):
```

Comment:
This is part of the reason why I think using `r` for the state, and `q` for the auxiliary variable is potentially confusing. Please update these :)

Reply:
Yeah, q here is just a local function variable and I used q for both the r and q flux functions. It would be good to update the local variable to r those that relate to r to make it more readable. Will do